sympy/polys/rootoftools.py

```diff
@@ -29,22 +29,98 @@
 from sympy.polys.domains import QQ
 
 from mpmath import mpf, mpc, findroot, workprec
-from mpmath.libmp.libmpf import prec_to_dps
+from mpmath.libmp.libmpf import dps_to_prec, prec_to_dps
 
 from sympy.utilities import lambdify, public, sift
 
-from sympy.core.compatibility import range
+from sympy.core.compatibility import range, ordered
 
 from math import log as mathlog
 
 __all__ = ['CRootOf']
 
-def _ispow2(i):
-    v = mathlog(i, 2)
-    return v == int(v)
 
-_reals_cache = {}
-_complexes_cache = {}
+
+class _pure_key_dict(dict):
```

Comment:
`object` would suffice.

Reply:
I added the `__contains__` method.

Reply:
Can't you just use a different key, common both for Poly and PurePoly?